### PR TITLE
New module @turf/clusters

### PR DESCRIPTION
**Pull request by stebogit**
First draft of `@turf/cluster`, Ref: #33 

```javascript
var clusters = require('@turf/clusters')
var points = // featurecollection of points
var numberOfClusters = 4
var clustered = clusters(points, numberOfClusters)

// clustered is an object with the following attributes:
// - `points`: the input (featurecollection of) points with each point given a `cluster` number property
// - `centroids`: featurecollection of centroid points, each with its `cluster` number as property
```
screen shot 2017-06-13 at 1 22 42 am

Uses [`clusters`](https://github.com/NathanEpstein/clusters) for simplicity, but there might be a faster/better implementation.
Not sure if the output is ok, I tried to follow @tmcw's [suggestion](https://github.com/Turfjs/turf/issues/33#issuecomment-264036583).

Open to suggestions and further testing.

### CODE REVIEW

**Comment by rowanwins** [comment]
ooo this looks nifty @stebogit

**Comment by DenisCarriere** [comment]
- [ ] Change the module name to `cluster` => so far all the TurfJS module names are singular.
- [ ] 50% agree with the output type, at the moment I can't really think of a better way to do it 😕

**Comment by DenisCarriere** [comment]
- [ ] All `properties` are lost, when creating these clusters it would be useful to preserve the existing properties.

**Comment by stebogit** [comment]
@DenisCarriere I replaced `clusters` with [`skmeans`](https://github.com/solzimer/skmeans) which is incredibly faster (like thousands times!) and allows to conserve the input and all its properties.

**Comment by stebogit** [comment]
@DenisCarriere 
> so far all the TurfJS module names are singular.

The more I think about it the more I believe this should be called 'clusters', since it's returning (and really expected to return) multiple clusters. The majority of Turf modules are singular because they deal with or generate a singular "thing", but few of them do return multiple elements and they have plural names, like `isobands`, `isolines`, `kinks`, `polygon-tangents`, `helpers`(functions).

**Comment by DenisCarriere** [comment]
👍 Ok that works, we can name it `clusters`.

**Comment by DenisCarriere** [comment]
>  I replaced clusters dependency with skmeans which allows to compute incredibly faster (like thousands times!) and to conserve the input and all its properties.

That's great, yea it was really slow before... I'm glad it's a lot faster now

File: packages/turf-cluster/index.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+ */
+module.exports = function (points, numberOfClusters) {
+    // Input validation
+    collectionOf(points, 'Point', 'Input must contain Points');
+    // Default Params
+    var count = points.features.length;
+    numberOfClusters = numberOfClusters || Math.round(Math.sqrt(count / 2));
+
+    // collect points coordinates
+    var data = featureReduce(points, function (prevValue, currentFeature) {
```
Comment (DenisCarriere) [inline packages/turf-cluster/index.js]: Using `coordEach` makes it about ~30-80% faster.

```javascript
    // collect points coordinates
    var data = [];
    coordEach(points, function (coord) {
        data.push(coord);
    });
```

**Comment by DenisCarriere** [comment]
- [ ] Number of points must be greater or equal to `numberOfClusters` (add as input validation)

File: packages/turf-cluster/index.js (hunk 1 above)
Comment (stebogit) [inline packages/turf-cluster/index.js]: 👍  good to know! I still don't master these methods 😕

**Comment by stebogit** [comment]
@DenisCarriere I added the validation on `numberOfClusters`

**Comment by DenisCarriere** [comment]
@stebogit 👍 Looks good!

**Comment by cyrilchapon** [comment]
Awesome

**Comment by DenisCarriere** [comment]
We should be pushing a new minor release soon with all these new modules being merged in the past week.

**Comment by cyrilchapon** [comment]
What about the complexity on adding a `maxDistance` factor, as described in #33 ?

**Comment by stebogit** [comment]
@cyrilchapon I'm not an expert in clustering, even less in clustering algorithms, however it seems to me an algo requiring a `maxDistance` would be substantially different that the `k-means` one.
Even reading through #33 I don't quite get what kind of feature you're suggesting; could you please give an example of use and application for that clustering method?

**Comment by cyrilchapon** [comment]
@stebogit yep.. I'm not an expert too in custering algos, but it seems indeed that it would be a very different algorithm.

Speaking of API, I would be suggesting something like

```
var clusters = require('@turf/clusters')
var points = // featurecollection of points

//kmeans strategy
var options = {
  numberOfClusters: 4
}
var 
var kmeansClustered = clusters(clusters.consts.algorithms.KMEANS, points, options)

//OR

//maxDistance strategy
var options = {
  maxDistance: 3,
  maxDistanceUnit: 'km'
}
var kmeansClustered = clusters(clusters.consts.algorithms.SOMENAME, points, options)
```

or maybe

```
var clusters = require('@turf/clusters')
var points = // featurecollection of points

//kmeans strategy
var numberOfClusters = 4
var kmeansClustered = clusters.kmeans(points, numberOfClusters)

//OR

//maxDistance strategy
var maxDistance = 3
var maxDistanceUnit = 'km'
var kmeansClustered = clusters.somename(points, maxDistance, maxDistanceUnit)
```

But I'm not able to implement such a thing

**Comment by DenisCarriere** [comment]
Both strategies shouldn't be implemented in the same module, they should be separate modules with unique names.

A good to read blog post from @mourner https://www.mapbox.com/blog/supercluster/ about Clustering.

`Supercluster` isn't meant for TurfJS, however some of the source code is good to look at for inspiration on how to tackle the **distance** strategy.

Another good library built by @mourner is [`kdbush`](https://github.com/mourner/kdbush) which is used by `Supercluster` and this would most likely be needed in this new cluster library.

**Pull request merged**